crates/astria-conductor/src/conductor.rs

```diff
@@ -77,20 +87,145 @@ impl Future for Handle {
+                                RestartToken::Restart => self.restart(),
+                                RestartToken::Shutdown => break,
+                            }
+                        },
+                        Ok(Err(err)) => {
+                            let error = err.wrap_err("conductor task exited unexpectedly");
+                            Err(error)?;
+                        },
+                        Err(err) => {
+                            let error = eyre::ErrReport::from(err).wrap_err("conductor failed during shutdown");
+                            Err(error)?;
+                        }
+                    }
+                }
+            }
+        }
+        self.shutdown().await?;
+        Ok(())
+    }
+
+    /// Creates and spawns a new [`InnerConductorTask`] task with the same configuration, replacing
+    /// the previous one. This function should only be called after a graceful shutdown of the
+    /// inner conductor task.
+    fn restart(&mut self) {
+        info!("restarting conductor");
+        let new_handle = InnerConductorTask::new(self.cfg.clone(), self.metrics)
+            .expect("failed to create new conductor after restart")
+            .spawn();
+        self.conductor_inner_shutdown = new_handle.shutdown_token.clone();
+        self.handle = new_handle;
+    }
+
+    /// Initiates shutdown of all conductor tasks from the top down, ignoring a restart signal.
+    async fn shutdown(self) -> eyre::Result<()> {
+        self.conductor_inner_shutdown.cancel();
+        let shutdown_result = self.handle.await?;
+        shutdown_result?;
+        Ok(())
+    }
+
+    #[must_use]
+    pub fn spawn(self) -> ConductorHandle {
+        let shutdown = self.shutdown.clone();
+        let task = tokio::spawn(self.run_until_stopped());
+        ConductorHandle {
+            shutdown,
+            task: Some(task),
+        }
+    }
+}
+
+pin_project! {
+    /// A handle returned by [`InnerConductorTask::spawn`].
+    pub struct ConductorInnerHandle {
+        shutdown_token: CancellationToken,
+        task: Option<tokio::task::JoinHandle<eyre::Result<RestartToken>>>,
+    }
+}
+
+impl Future for ConductorInnerHandle {
+    type Output = Result<eyre::Result<RestartToken>, tokio::task::JoinError>;
+
+    fn poll(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        use futures::future::FutureExt as _;
+        let this = self.project();
+        let task = this
+            .task
+            .as_mut()
+            .expect("the Conductor handle must not be polled after shutdown");
+        task.poll_unpin(cx)
+    }
+}
+
+pub struct InnerConductorTask {
+    /// Token to signal to all tasks to shut down gracefully.
+    shutdown: CancellationToken,
+
     /// The different long-running tasks that make up the conductor;
     tasks: JoinMap<&'static str, eyre::Result<()>>,
 }
 
-impl Conductor {
+impl InnerConductorTask {
```

Comment:
I chose `ConductorInner` since I thought it was slightly more appropriate, and also to my mind `Impl` kind of implies a trait. But, if using `{MODULENAME}Impl` is standard I'm happy to change it

Reply:
It's common to use `Thing` in the public API, and `ThingImpl` (or `ThingInner`) for the implementation deatils.

I would have been ok with `ConductorInner` as I said :D I just took offense to using `InnerConductorTask`